### PR TITLE
Prevent borrowing when preemption could help

### PR DESCRIPTION
**Pull request by alculquicondor**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Calculate flavor assignments that could be satisfied with preemption.

If we find such assignment, prevent borrowing in the cohort. Actual preemption will be implemented in a follow up.

#### Which issue(s) this PR fixes:

Part of #83 

This also solves what #403 was trying to solve.

### CODE REVIEW

File: pkg/scheduler/flavorassigner/flavorassigner.go [hunk 1]
```diff
@@ -296,38 +419,52 @@ func flavorSelector(spec *corev1.PodSpec, allowedKeys sets.String) nodeaffinity.
...
 	var status Status
 	used := cq.UsedResources[rName][flavor.Name]
+	mode := NoFit
+	if val <= flavor.Min {
+		// The request can be satisfied by the min quota, if all active workloads
+		// in the ClusterQueue are preempted.
+		mode = ClusterQueuePreempt
+	}
+	borrow := used + val - flavor.Min
+	if borrow <= 0 {
```
Comment (mimowo) [inline pkg/scheduler/flavorassigner/flavorassigner.go:435]: It seems to me that readability could be improved by moving this `if` into the `if val <= flavor.Min {` branch to avoid mutating `mode` and to make it clear that this happens only if `val <= flavor.Min`. For example, due to the mutability of `mode` it is not obvious if the comment under `val <= flavor.Min`  describes the entering of the branch or the semantics of the mode `ClusterQueuePreempt`  - because the semantics of the mode depends on what happens later.

**Review by mimowo — COMMENTED** [review]
Major comments:
- IIUC when there is a workload requesting more than `>flavor.Min`, but `<flavor.Max` then we don't attempt preemption (neither passive nor active) to make room for it (either `Fit` or `NoFit` mode is returned). IIUC, I suggest to make it clear in the comments and PR description or support.
- I find the description of the modes hard to understand, suggested rephrasing
- I find the logic in `fitsFlavorLimits` hard to follow. Suggested avoiding mutation of `mode` to improve it.

File: pkg/scheduler/flavorassigner/flavorassigner.go [hunk 2]
```diff
@@ -296,38 +419,52 @@ func flavorSelector(spec *corev1.PodSpec, allowedKeys sets.String) nodeaffinity.
...
+		borrow = 0
+		// The request can be satisfied by the min quota, if all active workloads
+		// from other ClusterQueues in the cohort are preempted.
+		mode = CohortReclaim
+	}
 	if flavor.Max != nil && used+val > *flavor.Max {
 		status.append(fmt.Sprintf("borrowing limit for %s flavor %s exceeded", rName, flavor.Name))
-		return 0, &status
+		return mode, 0, &status
```
Comment (ahg-g) [inline pkg/scheduler/flavorassigner/flavorassigner.go:443]: this if statement will never be entered if we entered the previous one, perhaps hook them via an else to emphasize that we are now looking at whether we can borrow resources:

`} else if flavor.Max != nil && used+val > *flavor.Max {`

File: pkg/scheduler/flavorassigner/flavorassigner.go [hunk 3]
```diff
@@ -296,38 +419,52 @@ func flavorSelector(spec *corev1.PodSpec, allowedKeys sets.String) nodeaffinity.
...
 	}
 	cohortUsed := used
 	cohortTotal := flavor.Min
```
Comment (ahg-g) [inline pkg/scheduler/flavorassigner/flavorassigner.go:446]: can we name it `cohortAvailable`

File: pkg/scheduler/flavorassigner/flavorassigner.go [hunk 4]
```diff
@@ -296,38 +419,52 @@ func flavorSelector(spec *corev1.PodSpec, allowedKeys sets.String) nodeaffinity.
...
-	}
 
 	lack := cohortUsed + val - cohortTotal
-	if lack > 0 {
-		lackQuantity := workload.ResourceQuantity(rName, lack)
-		if cq.Cohort == nil {
-			status.append(fmt.Sprintf("insufficient quota for %s flavor %s, %s more needed", rName, flavor.Name, &lackQuantity))
+	if lack <= 0 {
+		return Fit, borrow, nil
+	}
```
Comment (ahg-g) [inline pkg/scheduler/flavorassigner/flavorassigner.go:455]: I feel the function will be more readable if we move this check and early exit to be done first thing in the function (first, we check if we fit without preemption).

File: pkg/scheduler/flavorassigner/flavorassigner.go [hunk 5]
```diff
@@ -296,38 +419,52 @@ func flavorSelector(spec *corev1.PodSpec, allowedKeys sets.String) nodeaffinity.
 	return nodeaffinity.GetRequiredNodeAffinity(&corev1.Pod{Spec: specCopy})
 }
 
-// fitsFlavorLimits returns whether a requested resource fits in a specific flavor's quota limits.
+// fitsFlavorLimits returns how could this flavor be assigned, according to the
+// remaining quota in the ClusterQueue and cohort.
 // If it fits, also returns any borrowing required.
-func fitsFlavorLimits(rName corev1.ResourceName, val int64, cq *cache.ClusterQueue, flavor *cache.FlavorLimits) (int64, *Status) {
+func fitsFlavorLimits(rName corev1.ResourceName, val int64, cq *cache.ClusterQueue, flavor *cache.FlavorLimits) (FlavorAssignmentMode, int64, *Status) {
```
[the remark this answers is not in the document]
Comment (ahg-g) [reply pkg/scheduler/flavorassigner/flavorassigner.go:425]: That is a good point, I think we want to support this for higher priority jobs.

File: pkg/scheduler/flavorassigner/flavorassigner.go [hunk 6]
```diff
@@ -219,35 +331,46 @@ func (a *Assignment) findFlavorForCodepResources(
...
+				Mode:   mode,
 				borrow: borrow,
 			}
 		}
-		if len(assignments) == len(requests) {
-			return assignments, nil
+
+		if worstFlavorMode > bestAssignmentMode {
+			bestAssignment = assignments
+			bestAssignmentMode = worstFlavorMode
```
Comment (ahg-g) [inline pkg/scheduler/flavorassigner/flavorassigner.go:366]: this is a unintuitive, assigning the worst as the best :); perhaps rename `bestAssignmentMode` to `selectedMode`

**Review by ahg-g — COMMENTED** [review]
We need to document the placement order, something like:

1) If it fits within quota
2) borrowing
3) reclaiming lent quota
4) preempting within the CQ based on priority

and discuss cases that we don't currently support:
1) preempting to borrow
what else?

File: pkg/scheduler/flavorassigner/flavorassigner.go [hunk 7]
```diff
@@ -104,19 +175,44 @@ func (psa *PodSetAssignment) toAPI() kueue.PodSetFlavors {
 
 type FlavorAssignmentMode int
 
+// The flavor assigment modes below are ordered from lowest to highest
+// preference.
 const (
-	// CohortFit means that there are enough unused resources in the cohort to
+	// NoFit means that there is no enough quota left to assign this flavor.
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/scheduler/flavorassigner/flavorassigner.go]: Done

File: pkg/scheduler/flavorassigner/flavorassigner.go [hunk 8]
```diff
@@ -40,12 +43,56 @@ type Assignment struct {
...
 
 func (a *Assignment) Borrows() bool {
 	return len(a.TotalBorrow) > 0
 }
 
+func (a *Assignment) RepresentativeMode() FlavorAssignmentMode {
+	if a.repMode != nil {
+		return *a.repMode
+	}
+	if len(a.PodSets) == 0 {
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/scheduler/flavorassigner/flavorassigner.go]: Actually, this was needed to identify when there was no assignment calculated. This is covered by scheduler tests.

File: pkg/scheduler/flavorassigner/flavorassigner.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/scheduler/flavorassigner/flavorassigner.go:443]: I see, this sentence matches my case "The proposed policies for preemption within cohort require that the Workload fits within the min quota of the ClusterQueue.". 

Although I'm a bit confused "In other words, we don't try to borrow quota when preempting.". It is possible (but maybe unlikely) that the big high priority workload gets starved and cannot borrow while the smaller workloads don't trigger preemption cause they are constantly fitting inside their ClusterQueues.

[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/scheduler/flavorassigner/flavorassigner.go:443]: yes, but if a single high priority workload uses more than the entire min quota for the ClusterQueue, you are probably asking for trouble :)

File: pkg/scheduler/flavorassigner/flavorassigner.go [hunk 9]
```diff
@@ -138,17 +138,24 @@ func (s *Status) Equal(o *Status) bool {
 	}))
 }
 
+// PodSetAssignment holds the assigned flavors and status messages for each of
+// the resources that the podset requests. Each assigned flavor is accompanied
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/scheduler/flavorassigner/flavorassigner.go]: :sweat_smile: 

using "pod set" now.

File: pkg/scheduler/flavorassigner/flavorassigner.go [hunk 10]
```diff
@@ -138,17 +138,24 @@ func (s *Status) Equal(o *Status) bool {
...
+// with an AssignmentMode.
+// Empty flavors can be interpreted as NoFit mode for all the resources.
+// Empty status can be interpreted as Fit mode for all the resources.
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/scheduler/flavorassigner/flavorassigner.go]: Done

File: pkg/scheduler/flavorassigner/flavorassigner.go [hunk 11]
```diff
@@ -173,22 +180,25 @@ func (psa *PodSetAssignment) toAPI() kueue.PodSetFlavors {
...
 const (
-	// NoFit means that there is no enough quota left to assign this flavor.
+	// NoFit means that there is not enough quota to assign this flavor.
 	NoFit FlavorAssignmentMode = iota
-	// ClusterQueuePreempt means that there is not enough min quota in the
-	// ClusterQueue, so we need to preempt other workloads in the ClusterQueue
-	// with lower priority.
+	// ClusterQueuePreempt means that there is not enough unused min quota in the
+	// ClusterQueue. Preempting other workloads in the ClusterQueue or waiting for
+	// them to finish might allow to assign this flavor.
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/scheduler/flavorassigner/flavorassigner.go]: Done

File: pkg/scheduler/flavorassigner/flavorassigner.go (hunk 1 above)
Comment (alculquicondor) [reply pkg/scheduler/flavorassigner/flavorassigner.go]: I'm happy with the current flow. It might make it easier to split each mode into it's own independent function in the future, if necessary.

**Comment by alculquicondor** [comment]
squashed into two commits

**Comment by ahg-g** [comment]
/lgtm
/hold

in case Michal has other comments

File: pkg/scheduler/flavorassigner/flavorassigner.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/scheduler/flavorassigner/flavorassigner.go:445]: agree, wondering if there is a good place in the docs to explain that, but maybe it can wait for feedback from users if it is a real scenario

File: pkg/scheduler/flavorassigner/flavorassigner.go (hunk 1 above)
Comment (mimowo) [reply pkg/scheduler/flavorassigner/flavorassigner.go]: I'm also happy with that approach

**Comment by mimowo** [comment]
LGTM
/hold cancel
